creator-node/src/routes/userReplicaSet.js

```diff
@@ -0,0 +1,197 @@
+const axios = require("axios")
+const { promisify } = require('util')
+const crypto = require('crypto')
+const randomBytes = promisify(crypto.randomBytes)
+
+const {
+  handleResponse,
+  errorResponseBadRequest,
+  successResponse,
+  parseCNodeResponse,
+  errorResponseServerError
+} = require("../apiHelpers")
+const { logger } = require('../logging')
+const config = require('../config')
+const { generateTimestampAndSignature, recoverWallet } = require("../apiSigning")
+
+module.exports = function (app) {
+  /**
+   * TODO
+   * - rate limit
+   * - consider also validating L1 healthy state eg valid bounds
+   * - move to component pattern
+   */
+
+  /**
+   * This route is called by an requesting node attempting to self register on L2 UserReplicaSetManager
+   *    contract (URSM) via Chain of Trust. The requesting node is requesting a signature from this node
+   *    (the proposer node) to submit as part of the contract registration transaction.
+   *
+   * Steps:
+   *  1. Short circuit if node is already registered on URSM
+   *  2. Fetch node info from L1 ServiceProviderFactory for spID
+   *    a. Error if no L1 record found
+   *  3. Confirm request was signed by delegate owner wallet registered on L1 for spID, given request signature artifacts
+   *  4. Confirm health check returns healthy and response data matches on-chain data
+   *    a. Confirm health check response was signed by delegate owner wallet registered on L1 for spID
+   *  5. Generate & return proposal signature artifacts
+   */
+  app.get('/ursm_request_for_proposal', handleResponse(async (req, res, next) => {
+    const libs = req.app.get('audiusLibs')
+
+    let { spID, timestamp: reqTimestamp, signature: reqSignature } = req.query
+
+    if (!spID || !reqTimestamp || !reqSignature) {
+      return errorResponseBadRequest('Must provide all required query parameters: spID, timestamp, signature')
+    }
+
+    spID = parseInt(spID)
+
+    /**
+     * Short circuit if node already registered on URSM contract
+     */
+    const alreadyRegisteredOnURSM = !libs.Utils.isZeroAddress(
+      await libs.contracts.UserReplicaSetManagerClient.getContentNodeWallet(spID)
+    )
+    if (alreadyRegisteredOnURSM) {
+      return errorResponseBadRequest(
+        `Cannot submit proposal signature - Node with spID ${spID} already registered on UserReplicaSetManager contract`
+      )
+    }
+
+    /**
+     * Fetch node info from L1 ServiceProviderFactory for spID
+     */
+    const nodeSpInfoFromSPFactory = await libs.ethContracts.ServiceProviderFactoryClient.getServiceEndpointInfo(
+      libs.creatorNode.serviceTypeName,
+      spID
+    )
+    if (!nodeSpInfoFromSPFactory) {
```

Comment:
this will never actually return null, it will be all zero'd out - probably best to check more thoroughly here

Reply:
good call, fixed